misc(treemap): unused-bytes view mode

connorjclark opened the pull request:
ref #11254

https://lighthouse-1gxbe1yl8-googlechrome.vercel.app/gh-pages/treemap/?debug

This adds a "view modes" data structure, which defines how to interpret the data when walking `currentTreemapRoot`. `unused-bytes` is the first view mode, which highlights nodes with especially large amounts of unused bytes, and partitions the treemap based on percentage of total unused bytes on the page.

This only shows unused bytes based on javascript coverage, but there's no reason in the future that we couldn't determine unused bytes of images, css, etc. and plug them into this view mode.

![image](https://user-images.githubusercontent.com/4071474/109247780-1d9e6e80-77aa-11eb-9755-d93ae6cf9c21.png)

patrickhulce commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -110,13 +121,40 @@ class TreemapViewer {
+      subLabel: TreemapUtil.formatBytes(this.currentTreemapRoot.resourceBytes),
+    });
+
+    if (this.currentTreemapRoot.unusedBytes !== undefined) {
```
```suggestion
    // Add the unused bytes visualization only if the usage data is available.
    if (this.currentTreemapRoot.unusedBytes !== undefined) {
```

Is this right?

patrickhulce commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -110,13 +121,40 @@ class TreemapViewer {
+      TreemapUtil.walk(this.currentTreemapRoot, (node, path) => {
+        // Only highlight leaf nodes of a certain size.
+        if (node.children) return;
+        if (!node.unusedBytes || node.unusedBytes < 50 * 1024) return;
```
extract this 50KiB to a constant? feels like it will get buried otherwise :)

patrickhulce replied:
also should this account for relative usage some? like if 90% of a module 500kb dep is used, feels a bit weird to flag it as unused (though a bloat visualization should 😆 )

connorjclark replied:
Copied the constants from the unused-audit for this, but I also refactored things a bit...

Now, it will only highlight anything in a d1 (depth one) node if it is greater than some number of unused bytes. The audit uses a 20KiB threshold to ignore an entire bundle, so I did the same here. For a specific leaf node (a module in a bundle), only highlight if more than half of the bytes are unused.

Later PRs will add a pie graph/something similar to show how much of a node is unused.

patrickhulce commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -45,12 +45,23 @@ class TreemapViewer {
     this.el = el;
     this.getHueForKey = TreemapUtil.stableHasher(TreemapUtil.COLOR_HUES);
 
+    // TODO: make "DataSelector" to switch between different groups or specific d1 nodes
```
I'm not really sure what this means, should I?

patrickhulce commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -156,17 +194,20 @@ class TreemapViewer {
-    const size = node.resourceBytes;
-    const total = this.currentTreemapRoot.resourceBytes;
+    const partitionBy = this.currentViewMode.partitionBy || 'resourceBytes';
+    // @ts-expect-error
```
can we fix this with better types on`ViewMode`? `partitionBy: 'resourceBytes' | 'unusedBytes'`

patrickhulce commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -156,17 +194,20 @@ class TreemapViewer {
-    const total = this.currentTreemapRoot.resourceBytes;
+    const partitionBy = this.currentViewMode.partitionBy || 'resourceBytes';
+    // @ts-expect-error
+    const value = node[partitionBy];
```
leave this as `size` or `bytes` if we're still forcing the use of `formatBytes`?

patrickhulce commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -202,30 +251,49 @@ class TreemapViewer {
-    modesEl.addEventListener('click', () => {
+    inputEl.addEventListener('click', () => {
+      treemapViewer.currentViewMode = viewMode;
       treemapViewer.show();
```
so `show` is like `rerender`?

connorjclark replied:
combined into just a single `render` fn.

patrickhulce reviewed (commented):
code mostly LGTM just minor stuff 💯  

I had a few UX notes while I was playing around with it, but it's been a while since I remember talking about this so might have already been covered.

1. It takes a fair bit of time to orient oneself to a new treemap, identifying your bundles and what files you're looking at. When you click "unused" then, it was quite jarring to have the entire thing reshuffle orders, so I'm second guessing if that's the best way to interact with this information. IIRC, our rationale was that luminance intensity is pretty useless? But the use of white v. no color was pretty powerful already. My current thinking is that it might be a more powerful call to action if you get to see the unused modules in the context of your entire bundle for the easy wins. Right now it's weird that I'm looking at 100% unused bytes, so it feels like there's no priority.
2. The percentages in the module labels were confusing in the context of an unused bytes visualization. I get that it's a treemap and we're saying "hey this module's unused bytes were 2% of your overall unused bytes" but I don't think I care. I kinda want to know how much of the module itself was unused in that percentage.
3. The usage breakdown feels pretty unactionable when it's an entire bundle without a source map and there's quite a bit of visual space dedicated to them. I don't have a suggestion on how to fix yet, maybe we split them into source mapped and non-source mapped with less emphasis on the giant blobs?

connorjclark commented:
> It takes a fair bit of time to orient oneself to a new treemap, identifying your bundles and what files you're looking at. When you click "unused" then, it was quite jarring to have the entire thing reshuffle orders, so I'm second guessing if that's the best way to interact with this information. IIRC, our rationale was that luminance intensity is pretty useless? But I was surprised in practice how powerful the use of white v. no color was already. My current thinking is that it might be a more powerful call to action if you get to see the unused modules in the context of your entire bundle for the easy wins. Right now it's weird that I'm looking at 100% unused bytes, so it feels like there's no priority.

I removed the `unusedBytes` partition for this view mode and I agree, it is much less jarring. For purposes of quick assessment I think this is the better experience so let's do that. I'm not certain that a "partition by unused bytes" couldn't be useful to _someone_. For myself, if I'm tackling the problem "let's reduce how much unused JS we are shipping", I'd appreciate the ability to partition by the metric I'm trying to reduce. But that can be relegated to the settings module.

> The percentages in the module labels were confusing in the context of an unused bytes visualization. I get that it's a treemap and we're saying "hey this module's unused bytes were 2% of your overall unused bytes" but I don't think I care. I kinda want to know how much of the module itself was unused in that percentage.

Later PRs will add a gauge to each module to indicate how much is unused. also, the table component that is coming up soon offers another way to view this data. see gallery.googleplex.com/projects/MCHbtQVoQ2HCZeok9-CasvMb/files/MCFFTHkODMy6RyA6lXJuFGJb

> The usage breakdown feels pretty unactionable when it's an entire bundle without a source map and there's quite a bit of visual space dedicated to them. I don't have a suggestion on how to fix yet, maybe we split them into source mapped and non-source mapped with less emphasis on the giant blobs?

I don't want to deemphasize something just because it doesn't have a source map. Perhaps an indicator for these nodes that "A source map will provide more granular data" will suffice?

paulirish commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -156,17 +205,21 @@ class TreemapViewer {
+
+      // Only add label for bytes on the root node.
+      if (node === this.currentTreemapRoot) {
+        parts[1] = `${partitionBy}: ${parts[1]}`;
```
overwriting it feels weird. 

how about a thing like 

```js
if (bytes !== undefined && total !== undefined) {
	const str = node === this.currentTreemapRoot 
      ? ${partitionBy}: ${parts[1]}` 
      :  `${TreemapUtil.formatBytes(bytes)} (${Math.round(bytes / total * 100)}%)`;
    parts.push(str);
}
```

connorjclark commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -110,13 +124,48 @@ class TreemapViewer {
+      for (const d1Node of this.currentTreemapRoot.children || []) {
+        // Only highlight leaf nodes if entire node (ie a JS bundle) has greater than a certain
+        // number of unused bytes.
+        if (!d1Node.unusedBytes || d1Node.unusedBytes < UNUSED_BYTES_IGNORE_THRESHOLD) continue;
```
no, the treemap data sets unusedBytes straight from the coverage data, not the unused-javascript audit.

connorjclark commented on lighthouse-treemap/app/src/main.js:
```diff
@@ -191,6 +244,14 @@ class TreemapViewer {
       }
 
+      // A view can set nodes to highlight. If so, don't color anything else.
+      if (this.currentViewMode.highlightNodePaths) {
```
but, typescript?

paulirish commented on lighthouse-treemap/app/styles/treemap.css:
```diff
@@ -65,6 +65,12 @@ body {
   background-color: lightblue;
 }
 
+.view-mode__button {
```
btw you're losing your cursor styling because label has a default (who knew)

![image](https://user-images.githubusercontent.com/39191/111365472-c1639780-864f-11eb-8de1-eb96958940fb.png)

paulirish reviewed (approved):
lgtm. one small thing i noticed.
